developer/src/kmc-ldml/test/test-visual-keyboard-compiler.ts

```diff
@@ -0,0 +1,228 @@
+  this.slow(500); // 0.5 sec -- json schema validation takes a while
+
+  it('should build fixtures', async function() {
+    // Let's build basic.xml
+
+    // It should match basic.kvk (built from basic-kvk.txt)
+    const inputFilename = makePathToFixture('basic.xml');
+    const binaryFilename = makePathToFixture('basic-kvk.txt');
+
+    // Compile the visual keyboard
+    const k = new LdmlKeyboardCompiler();
+    assert.isTrue(await k.init(compilerTestCallbacks, {...compilerTestOptions, saveDebug: true, shouldAddCompilerVersion: false}));
+    const source = k.load(inputFilename);
+    checkMessages();
+    assert.isNotNull(source, 'k.load should not have returned null');
+
+    const valid = await k.validate(source);
+    checkMessages();
+    assert.isTrue(valid, 'k.validate should not have failed');
+
+    let kmx = await k.compile(source);
+    assert(kmx, 'k.compile should not have failed');
+
+    const vk = (new LdmlKeyboardVisualKeyboardCompiler(compilerTestCallbacks)).compile(kmx.kmxplus, path.basename(inputFilename, '.xml'));
+    checkMessages();
+    assert.isNotNull(vk, 'LdmlKeyboardVisualKeyboardCompiler.compile should not have returned null');
+
+    // Use the builder to generate the binary output file
+    const writer = new KvkFileWriter();
+    const code = writer.write(vk);
+    assert.isEmpty(compilerTestCallbacks.messages);
+    assert.isNotNull(code);
+
+    // Compare output
+    let expected = await hextobin(binaryFilename, undefined, {silent:true});
+
+    assert.deepEqual<Uint8Array>(code, expected);
+  });
+
+  it('should support various modifiers', async function() {
+    const xml = stripIndent`
+      <?xml version="1.0" encoding="UTF-8"?>
+      <keyboard3 xmlns="https://schemas.unicode.org/cldr/45/keyboard3" locale="mt" conformsTo="45">
+        <info name="minimal"/>
+        <keys>
+          <key id="x" output="y" />
+        </keys>
+        <layers formId="us">
+          <layer modifiers="none"><row keys="x" /></layer>
+          <layer modifiers="shift"><row keys="x" /></layer>
+          <layer modifiers="altR"><row keys="x" /></layer>
+          <layer modifiers="ctrlR altR, shift altR"><row keys="x" /></layer>
+          <layer modifiers="caps"><row keys="x" /></layer>
+        </layers>
+      </keyboard3>
+    `;
+
+    const vk = await loadVisualKeyboardFromXml(xml, 'test');
+
+    assert.equal(vk.keys.length, 5);
+    assert.equal(vk.keys[0].shift, VisualKeyboard.VisualKeyboardShiftState.KVKS_NORMAL);
+    assert.equal(vk.keys[1].shift, VisualKeyboard.VisualKeyboardShiftState.KVKS_SHIFT);
+    assert.equal(vk.keys[2].shift, VisualKeyboard.VisualKeyboardShiftState.KVKS_RALT);
+    assert.equal(vk.keys[3].shift, VisualKeyboard.VisualKeyboardShiftState.KVKS_RCTRL | VisualKeyboard.VisualKeyboardShiftState.KVKS_RALT);
+    assert.equal(vk.keys[4].shift, VisualKeyboard.VisualKeyboardShiftState.KVKS_SHIFT | VisualKeyboard.VisualKeyboardShiftState.KVKS_RALT);
```

Comment:
ok, so do these correspond to 
```xml
<layer modifiers="ctrlR altR, shift altR"><row keys="x" /></layer>
````

and then
```xml
<layer modifiers="caps"><row keys="x" /></layer>
```
has a .shift = null?

Reply:
Yes, the OSK does not support Caps so we omit the Caps layer.